### PR TITLE
replace future with celery.five Fixes #6250, and reraise to include outer tracebacks

### LINKED ISSUE #6250: dependency on future module breaks many things (including vim ultisnips)
the dependency on future-0.18.2 adds many packages that squat on py3 stdlib names eg:

```
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/_dummy_thread/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/_markupbase/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/_thread/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/builtins/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/copyreg/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/future-0.18.2.dist-info/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/future/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/html/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/http/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/libfuturize/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/libpasteurize/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/past/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/queue/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/reprlib/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/socketserver/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/tkinter/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/winreg/*
    /home/graingert/.virtualenvs/celery2/lib/python2.7/site-packages/xmlrpc/*
```

this breaks packages that rely on feature detection of these modules.

I propose replacing future with six because:

* future does not provide a wheel, six has a wheel
* future last release Oct 31, 2019, six released Oct 31, 2019
* future is 829.2 kB, six is (11.0 kB) 
* celery already depends on six, via kombu and importlib-metadata, and so swapping out future with six would save 829.2 kB, rather than just 818.2 kB

# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [ ] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [ ] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
```

</p>
</details>

# Expected Behavior

# Actual Behavior

### PR DESCRIPTION
**Pull request by graingert**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description
Fixes #6250, relates, 0463bff0530b8aef8d31bd8039f245735295db59

### CODE REVIEW

File: celery/backends/base.py [hunk 1]
```diff
@@ -443,6 +455,7 @@ def store_result(self, task_id, result, state,
                                    request=request, **kwargs)
                 return result
             except Exception as exc:
+                tb = sys.exc_info()[2]
```
Comment (graingert) [inline celery/backends/base.py:458]:
```suggestion
```

File: celery/backends/base.py [hunk 2]
```diff
@@ -454,7 +467,10 @@ def store_result(self, task_id, result, state,
                             self.max_sleep_between_retries_ms, True) / 1000
                         self._sleep(sleep_amount)
                     else:
-                        raise_with_traceback(BackendStoreError("failed to store result on the backend", task_id=task_id, state=state))
+                        _raise_with_traceback(
+                            BackendStoreError("failed to store result on the backend", task_id=task_id, state=state),
+                            tb,
```
Comment (graingert) [inline celery/backends/base.py:472]:
```suggestion
                            traceback,
```

File: celery/backends/base.py [hunk 3]
```diff
@@ -454,7 +466,10 @@ def store_result(self, task_id, result, state,
                             self.max_sleep_between_retries_ms, True) / 1000
                         self._sleep(sleep_amount)
                     else:
-                        raise_with_traceback(BackendStoreError("failed to store result on the backend", task_id=task_id, state=state))
+                        _raise_with_traceback(
+                            BackendStoreError("failed to store result on the backend", task_id=task_id, state=state),
+                            traceback,
```
Comment (graingert) [inline celery/backends/base.py:471]: @mchataigner is this the traceback you intend to raise with?

If you pass `None` in here, I think it's the same as the plain `raise BackendStoreError(...)`

File: celery/backends/base.py [hunk 4]
```diff
@@ -73,6 +72,19 @@ def unpickle_backend(cls, args, kwargs):
     return cls(*args, app=current_app._get_current_object(), **kwargs)
 
 
+
+if sys.version_info < (3, 0, 0):
+    from six import reraise as _reraise
```
Comment (auvipy) [inline celery/backends/base.py:77]: is it available in celery.five?

File: celery/backends/base.py [hunk 5]
```diff
@@ -532,7 +548,7 @@ def get_task_meta(self, task_id, cache=True):
                             self.max_sleep_between_retries_ms, True) / 1000
                         self._sleep(sleep_amount)
                     else:
-                        raise_with_traceback(BackendGetMetaError("failed to get meta", task_id=task_id))
+                        _raise_with_traceback(BackendGetMetaError("failed to get meta", task_id=task_id), tb)
```
Comment (graingert) [inline celery/backends/base.py:551]: @mchataigner is this `tb` the right traceback?

File: celery/backends/base.py (hunk 4 above)
Comment (graingert) [reply celery/backends/base.py:77]: ah yes, absolutely: https://github.com/celery/vine/blob/master/vine/five.py#L192

Comment (graingert) [reply celery/backends/base.py:77]: and you don't need any of this `if sys.version_info` !

**Comment by thedrow** [comment]
Why do we need future or six anyway?

**Comment by graingert** [comment]
@thedrow we don't! we can use five!

also for this code, the tracebacks didn't seem to be attached anyway - so plain raise would have worked fine.

I think the intention was to reraise with the original traceback.

**Comment by graingert** [comment]
@auvipy it's so much better with `celery.five` thanks!

**Comment by graingert** [comment]
@auvipy do you want me to replace the `reraise(C, C(v), tb)` with `raise C(v)`? or are you ok with my guess as to what the tb should have been?

File: celery/backends/base.py [hunk 6]
```diff
@@ -73,6 +72,7 @@ def unpickle_backend(cls, args, kwargs):
     return cls(*args, app=current_app._get_current_object(), **kwargs)
 
 
+
```
Comment (graingert) [inline celery/backends/base.py:75]:
```suggestion
```

**Comment by auvipy** [comment]
thanks a lot for your efforts!

**Comment by graingert** [comment]
@auvipy did I reraise the right tracebacks?

**Pull request merged**